File: src/functions/utils/global.ts
import { createEmbed } from "@magicyan/discord";
import { Canvas, loadImage } from "@napi-rs/canvas";
import {
    AttachmentBuilder,
    Events,
    GuildMember,
    PartialGuildMember,
    TextChannel,
    User,
} from "discord.js";
import { join } from "path";
import { channels, icon, roles } from "#functions";
import { settings } from "#settings";

export async function globalMessage( event: Events, member: GuildMember | PartialGuildMember | undefined, user: User, channel: TextChannel) {
    const canvas = new Canvas(2800, 560);
    const context = canvas.getContext("2d");
    
    let bgpath = join(__rootname, "assets/canvas/");
    if (event === Events.GuildMemberAdd && typeof member !== "undefined") {
        const { joinedTimestamp } = member;
        const accountAge = Date.now() - user.createdTimestamp;
        const joinAge = Date.now() - (joinedTimestamp || 0);
        const timeLimit = 60 * 1000;
        bgpath =
            joinAge < timeLimit && accountAge > timeLimit
                ? join(bgpath, "card-new-member.png")
                : join(bgpath, "card-back-to-server.png");
    } else {
        bgpath = join(bgpath, event === Events.GuildMemberRemove ? "card-left.png" : "card-mod.png");
    }
    const background = await loadImage(bgpath);
    context.drawImage(background, 0, 0, canvas.width, canvas.height);
    context.save();

    const avatar = await loadImage(user.displayAvatarURL({ size: 512 }));
    context.beginPath();
    context.arc(200, 200 + 160, 200, 0, Math.PI * 2);
    context.clip();
    context.drawImage(avatar, 0, 160, 400, 400);
    context.restore();

    const { username, displayName } = user;

    let usernameFontSize = 200;
    context.fillStyle = "#FFFFFF";
    do {
        context.font = `medium ${--usernameFontSize}px Fredoka`;
    } while (context.measureText(username).width > canvas.width - 400);
    context.textBaseline = "middle";
    context.fillText(username, 530, 140 + usernameFontSize / 2);

    let displayNameFontSize = 96;
    context.fillStyle = "#FFFFFF";
    do {
        context.font = `medium ${--displayNameFontSize}px Roboto`;
    } while (context.measureText(`@${displayName}`).width > canvas.width - 400);
    context.textBaseline = "middle";
    context.fillText(`@${displayName}`, 530, 380 + displayNameFontSize / 2);

    const buffer = await canvas.encode("png");
    const attachment = new AttachmentBuilder(buffer, { name: "card.png" });

    let utc  = "";
    if (event === Events.GuildMemberAdd && typeof member !== "undefined") {
        const { joinedTimestamp } = member;
        if (joinedTimestamp) {
            const timestamp = Math.floor(joinedTimestamp / 1000);
            utc = `<t:${timestamp}:F>`;
        }
    } else {
        const timestamp = Math.floor(Date.now() / 1000);
        utc = `<t:${timestamp}:F>`
    }
    
    channel.send({ content: utc, files: [attachment] });
}

export async function globalBoost(member: GuildMember) {
    const { guild, id, user } = member;
    const { globalName, username } = user;
    const guild_channel = await guild.channels.fetch(channels.announcement);
    if (!guild_channel) return;
    const avatarURL = user.avatarURL({size: 256});
    const embed = createEmbed({
        color: settings.colors.nitro,
        author: {
            name: globalName || username,
            iconURL: avatarURL || undefined
        },
        description: `**${icon.boost} <@${id}> became a <@&${roles.boosters}>**\n\n🚀 Thanks for boosting the server!`,
        thumbnail: avatarURL || undefined
    });
    const channel = await guild_channel.fetch();
    if (channel.isTextBased()) {
        channel.send({content: "||@everyone @here||", embeds: [embed]});
    }
}

File: src/utils/global.rs
use crate::discord::*;
use crate::utils::*;
use crate::data::*;
use crate::assets::*;
use serenity::all::{
    CacheHttp, ChannelId, Colour, Context, CreateAttachment, CreateEmbed, CreateEmbedAuthor,
    CreateMessage, GuildId, Member, User,
};
use skia_safe::{Data, EncodedImageFormat, ISize, Point};
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Eq, PartialEq)]
pub enum EventType {
    MemberAdded,
    MemberRemoved,
    BanAdded,
}

pub async fn global_message(
    ctx: &Context,
    channel_id: &ChannelId,
    event: EventType,
    member: Option<&Member>,
    user: &User,
) {
    // Fetch avatar
    let mut user_avatar: Vec<u8> = vec![];

    if let Some(avatar_hash) = user.avatar {
        let avatar_url = display_avatar_url(user.id.get(), &avatar_hash.to_string(), 512);
        if let Ok(res) = reqwest::get(avatar_url).await {
            if let Ok(bytes) = res.bytes().await {
                user_avatar = bytes.to_vec();
            }
        }
        if user_avatar.is_empty() {
            user_avatar = IMG_DEFAULT_AVATAR.to_vec();
        }
    } else {
        user_avatar = IMG_DEFAULT_AVATAR.to_vec();
    }

    let background = match event {
        EventType::MemberAdded => {
            let date = match SystemTime::now().duration_since(UNIX_EPOCH) {
                Ok(duration) => duration.as_millis(),
                Err(_) => 0,
            };
            let join_age = match member.unwrap().joined_at {
                Some(joined_at) => date - joined_at.timestamp_millis() as u128,
                None => 0,
            };
            let account_age = date - user.id.created_at().timestamp_millis() as u128;
            const TIME_LIMIT: u128 = 300 * 1000;
            if join_age < TIME_LIMIT {
                CARD_NEW
            } else if account_age < TIME_LIMIT {
                CARD_NEW
            } else {
                CARD_BACK
            }
        }
        EventType::MemberRemoved => CARD_LEFT,
        EventType::BanAdded => CARD_MOD,
    };

    let mut data = vec![];
    // Scope to fix safe issue with skia
    {
        let mut surface = match skia_safe::surfaces::raster_n32_premul(ISize {
            width: 2800,
            height: 560,
        }) {
            Some(surface) => surface,
            None => {
                error("Failed to create canvas surface.");
                return;
            }
        };
        let canvas = surface.canvas();

        let background_image = match load_image_from_bytes(background) {
            Some(image) => image,
            None => {
                error("Failed to load background image.");
                return;
            }
        };
        canvas.draw_image(&background_image, Point { x: 0.0, y: 0.0 }, None);
        canvas.save();

        // Avatar
        let cdn_avatar = match load_image_from_bytes(&user_avatar) {
            Some(image) => image,
            None => {
                error("Failed to decode user avatar image.");
                return;
            }
        };
        let avatar = match resize_image(cdn_avatar, 400, 400) {
            Some(image) => image,
            None => {
                error("Failed to resize user avatar image.");
                return;
            }
        };
        draw_circle(canvas, avatar, 204.0, 360.0, 200.0);

        if !draw_text_with_font(canvas, &user.name, FONT_FREDOKA, 200.0, 530.0, 140.0) {
            error("Failed to resize user avatar image.");
            return;
        }

        let nickname = match user.global_name.as_ref() {
            Some(nickname) => nickname,
            None => "Undefined",
        };
        if !draw_text_with_font(
            canvas,
            &format!("@{}", nickname),
            FONT_ROBOTO,
            96.0,
            530.0,
            380.0,
        ) {
            error("Failed to resize user avatar image.");
            return;
        }

        let image = surface.image_snapshot();
        let encoded_data = match image.encode(None, EncodedImageFormat::PNG, Some(100)) {
            Some(data) => data,
            None => {
                error("Failed to encode card image.");
                return;
            }
        };

        data = encoded_data.to_vec();
    }

    let mut utc = String::new();
    if event == EventType::MemberAdded {
        let joined_at = member.unwrap().joined_at.unwrap_or_default().timestamp();
        utc.push_str(&format!("<t:{}:F>", joined_at));
    }
    let attachment = CreateAttachment::bytes(data, "card.png");
    let message = CreateMessage::new()
        .content(utc)
        .add_files(vec![attachment]);

    if let Err(err) = channel_id.widen().send_message(ctx.http(), message).await {
        error(&format!(
            "Error trying to send card to system channel\nʟ {:?}",
            err
        ));
    }
}

pub async fn global_boost(ctx: &Context, user: &User, guild_id: &GuildId) {
    let color = Colour::new(str_hex_to_u32(&CONSTANTS.colors.nitro));
    let avatar_url = user.avatar_url().unwrap_or_default();
    let username = user.global_name.clone().unwrap_or(user.name.clone());
    let description = format!(
        "**<a:boost:{}> <@${}> became a <@&${}>**\n\n🚀 Thanks for boosting the server!",
        &EMOJIS.animated.boost, user.id, &GUILD.roles.boosters
    );

    let author = CreateEmbedAuthor::new(username.as_str()).icon_url(&avatar_url);
    let embed = CreateEmbed::new()
        .color(color)
        .author(author)
        .description(description)
        .thumbnail(&avatar_url);

    let channel = match guild_id.channels(ctx.http()).await {
        Ok(channels) => {
            let id = ChannelId::new(str_to_u64(&GUILD.channels.announcement));
            if let Some(channel) = channels.get(&id).cloned() {
                channel
            } else {
                error(&format!("Guild channel not found!"));
                return;
            }
        }
        Err(err) => {
            error(&format!("Failed to remove member role!\n└ {:?}", err));
            return;
        }
    };

    let payload = CreateMessage::new()
        .content("||@everyone @here||")
        .embed(embed);
    if let Err(err) = channel.send_message(ctx.http(), payload).await {
        error(&format!("Failed to send message!\n└ {:?}", err));
    }
}
